### PR TITLE
driver: sensor: Add support for BMP388 pressure sensor

### PR DESCRIPTION
**Pull request by xodus7**

### CODE REVIEW

**Comment by xodus7** [comment]
I just noticed this PR #22918 from Feb. It's labeled stale and hasn't had any recent activity. Is the core project not accepting new sensor drivers? I'm happy to cancel the PR if it isn't or the original PR is preferred.

File: drivers/sensor/bmp388/Kconfig
Comment (mbolivar-nordic) [inline drivers/sensor/bmp388/Kconfig:137]: This all looks like it belongs in DT.

File: drivers/sensor/bmp388/bmp388.c
Comment (mbolivar-nordic) [inline drivers/sensor/bmp388/bmp388.c:54]: Same buffers for TX and RX? Are you sure? I can only find one other instance of that in the tree, added in 810920f8d553717f7175fd8e73e571a41f136d8e. @tbursztyka are the semantics here well-defined?

File: drivers/sensor/bmp388/bmp388.c
Comment (mbolivar-nordic) [inline drivers/sensor/bmp388/bmp388.c:691]: I think this driver should be made multi-instance.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
> Is the core project not accepting new sensor drivers? I'm happy to cancel the PR if it isn't or the original PR is preferred.

There just aren't enough reviewers to go around.

Overall I think this seems reasonable, but I do have a few comments and questions.

File: drivers/sensor/bmp388/bmp388.c
Comment (xodus7) [reply drivers/sensor/bmp388/bmp388.c:691]: I took a pass at this a while back but the ```#if``` blocks in the definition of the ```bmp388_config``` struct make it hard to wrap this all in a macro. Many sensors drivers are currently single instance, is this a hard requirement?

Comment (mbolivar-nordic) [reply drivers/sensor/bmp388/bmp388.c:691]: > Many sensors drivers are currently single instance, is this a hard requirement?

That's up to @MaureenHelm as the sensor maintainer.

File: drivers/sensor/bmp388/Kconfig
Comment (avisconti) [reply drivers/sensor/bmp388/Kconfig:137]: I think that at least odr frequency should be put in DT.

File: drivers/sensor/bmp388/bmp388.c
Comment (avisconti) [reply drivers/sensor/bmp388/bmp388.c:691]: We are slightly moving all sensors, or at least the most recent one, to be multi-instance. I would say that this should be the rule for every new sensor implementation. I leave to @mhelm the last word on it.

**Review by avisconti — CHANGES_REQUESTED** [review]

File: drivers/sensor/bmp388/bmp388.c
Comment (xodus7) [reply drivers/sensor/bmp388/bmp388.c]: Got it, I've push a change that makes this multi-instance aware.

File: drivers/sensor/bmp388/bmp388.c
Comment (xodus7) [reply drivers/sensor/bmp388/bmp388.c:54]: The BMI160 driver uses the same buffers as well. This is a common thing to do with SPI.

File: drivers/sensor/bmp388/Kconfig
Comment (xodus7) [reply drivers/sensor/bmp388/Kconfig]: I've made these all DT properties as requested.

File: drivers/sensor/bmp388/bmp388.c
Comment (mbolivar-nordic) [reply drivers/sensor/bmp388/bmp388.c:56]: > The BMI160 driver uses the same buffers as well.

The same `spi_buf`, yes, but not the same `spi_buf_set` as is being done here.

I'd still like a clarification from @tbursztyka on whether using the same `spi_buf_set` for both TX and RX allowed by the SPI API. Would be good to get it documented one way or the other.

File: drivers/sensor/bmp388/bmp388.c
Comment (mbolivar-nordic) [inline drivers/sensor/bmp388/bmp388.c:560]: I think this should get updated to use DEVICE_DT_GET to get the bus pointer into the configuration structure in ROM at build time, then use device_is_ready() to check that it's initialized properly here, as is being done to existing drivers throughout the tree for 2.6.

File: drivers/sensor/bmp388/bmp388.c
Comment (MaureenHelm) [inline drivers/sensor/bmp388/bmp388.c:70]: It's an unlikely configuration, but this won't work if there's one i2c instance and another spi instance. Most other i2c/spi sensor drivers abstract this with a per-instance function pointer to access the sensor registers.

File: drivers/sensor/bmp388/bmp388.c
Comment (MaureenHelm) [inline drivers/sensor/bmp388/bmp388.c:413]: can you document where these constants come from?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please add this sensor to `tests/drivers/build_all/sensors_a_h.conf` to ensure it gets built in CI.

>> Is the core project not accepting new sensor drivers? I'm happy to cancel the PR if it isn't or the original PR is preferred.

> There just aren't enough reviewers to go around.

This is indeed the problem. I'm sorry for the delay in my review.

File: drivers/sensor/bmp388/bmp388.c
Comment (xodus7) [reply drivers/sensor/bmp388/bmp388.c]: I made a pass at this, however, I'm not able to test all the various combinations :-)

File: drivers/sensor/bmp388/bmp388.h
Comment (avisconti) [inline drivers/sensor/bmp388/bmp388.h:144]: I was just suggested by @mbolivar-nordic to not use union here, just to avoid to lose a lot of space in I2C case. See (if you want) what I have done instead in #33506 (two separate structures)

File: drivers/sensor/bmp388/bmp388.c
Comment (avisconti) [inline drivers/sensor/bmp388/bmp388.c:735]: The cs part should be also treated wih SPI_CONFIG_DT_INST() (See below)

File: drivers/sensor/bmp388/bmp388.c
Comment (avisconti) [inline drivers/sensor/bmp388/bmp388.c:72]: Is this case correct? Not too many bytes skipped?

**Review by avisconti — CHANGES_REQUESTED** [review]
Very nice improvement! But I think that there are new DT helpers you may use. So, we are very close now to the final stage in my opinion.

File: drivers/sensor/bmp388/bmp388.c
Comment (avisconti) [inline drivers/sensor/bmp388/bmp388.c:751]: Use SPI_CONFIG_DT_INST(). (See #31775)

File: drivers/sensor/bmp388/bmp388.c
Comment (avisconti) [inline drivers/sensor/bmp388/bmp388.c:762]: USe GPIO_DT_SPEC_GET(). See #31775

File: drivers/sensor/bmp388/bmp388.c
Comment (xodus7) [reply drivers/sensor/bmp388/bmp388.c:72]: Yes, see comment above about dummy byte.

Comment (avisconti) [reply drivers/sensor/bmp388/bmp388.c:72]: So to read from SPI you perform a TX transaction (2 bytes) with reg + 1 DUMMY, and then again same TX transaction followed by RX transaction of length byte. Am I missing something about spi_transceive()?

File: drivers/sensor/bmp388/bmp388.h
Comment (avisconti) [reply drivers/sensor/bmp388/bmp388.h:144]: No, it is correct. In my case was different. Sorry for noise.

File: drivers/sensor/bmp388/bmp388.c
Comment (avisconti) [reply drivers/sensor/bmp388/bmp388.c:72]: To me this is the last point not clear (but if it works maybe I'm missing something). All the rest of open points seem good.

**Review by mbolivar-nordic — APPROVED** [review]
Binding and device instantiation look great. Thanks.

File: drivers/sensor/bmp388/bmp388.c
Comment (xodus7) [reply drivers/sensor/bmp388/bmp388.c:72]: TX/RX happen at the same time, so we write/read two bytes (we don't care about read bytes but use the same buffer for efficiency), then we read N bytes. This works in my STM32L4 based setup.

**Review by MaureenHelm — APPROVED** [review]

**Review by avisconti — APPROVED** [review]

**Comment by xodus7** [comment]
Thanks for the extensive review everyone!

**Pull request merged**
